common/src/main/scala/org/apache/spark/sql/comet/execution/arrow/CometArrowConverters.scala

```diff
@@ -0,0 +1,117 @@
+
+package org.apache.spark.sql.comet.execution.arrow
+
+import scala.collection.JavaConverters.asScalaBufferConverter
+
+import org.apache.arrow.vector.VectorSchemaRoot
+import org.apache.spark.TaskContext
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.execution.arrow.ArrowWriter
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.ArrowUtils
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+import org.apache.comet.vector.CometVector
+
+object CometArrowConverters extends Logging {
+  // this is similar how Spark converts internal row to Arrow format except that we are transforming
+  // the result batch to Comet's Internal ColumnarBatch instead of serialized bytes.
+  private[sql] class ArrowBatchIterator(
+      rowIter: Iterator[InternalRow],
+      schema: StructType,
+      maxRecordsPerBatch: Long,
+      timeZoneId: String,
+      context: TaskContext)
+      extends Iterator[ColumnarBatch]
+      with AutoCloseable {
+
+    // todo: hmm, we need to handle arrow shading problem, as Spark may use a different version of
+    //    arrow.
+    private val arrowSchema =
+      ArrowUtils.toArrowSchema(schema, timeZoneId)
+    // Reuse the same root allocator here, maybe we should also reuse the same allocator in the
+    // comet code base
+    private val allocator =
+      ArrowUtils.rootAllocator.newChildAllocator(
+        s"to${this.getClass.getSimpleName}",
+        0,
+        Long.MaxValue)
+
+    private val root = VectorSchemaRoot.create(arrowSchema, allocator)
+    private val arrowWriter = ArrowWriter.create(root)
+    private var columnarBatch: ColumnarBatch = null
+
+    Option(context).foreach {
+      _.addTaskCompletionListener[Unit] { _ =>
+        close()
+      }
+    }
+
+    override def hasNext: Boolean = rowIter.hasNext || {
+      close()
+      false
+    }
+
+    override def next(): ColumnarBatch = {
+      if (columnarBatch != null) {
+        // reset the arrowWrite and columnarBatch. The reset method is called only after the
+        // columnarBatch is consumed by the caller.
+        arrowWriter.reset()
+        columnarBatch = null
+      }
+      var rowCount = 0L
+      while (rowIter.hasNext && (maxRecordsPerBatch <= 0 || rowCount < maxRecordsPerBatch)) {
+        val row = rowIter.next()
+        arrowWriter.write(row)
+        rowCount += 1
+      }
+      arrowWriter.finish()
+      columnarBatch = wrapperFor(root)
+      columnarBatch
+    }
+
+    override def close(): Unit = {
+      if (columnarBatch != null) {
+        arrowWriter.reset()
+        columnarBatch = null
+      }
+      root.close()
+      allocator.close()
+    }
+  }
+
+  private def wrapperFor(root: VectorSchemaRoot) = {
+    val columns = root.getFieldVectors.asScala.map { vector =>
+      CometVector.getVector(vector, false)
+    }
+    new ColumnarBatch(columns.toArray, root.getRowCount)
+  }
+
+  def toArrowBatchIterator(
```

Comment:
Sorry, I missed your earlier message. Yes, I agree that the first approach is more straightforward to implement. Regarding memory management, it would be nice if we can somehow hook it with the `TaskMemoryManager` since the row to columnar conversion does require non-trivial amount of memory.

Will take a look at the PR soon.